packages/material-ui/src/Menu/Menu.js

```diff
@@ -71,13 +104,33 @@ const Menu = React.forwardRef(function Menu(props, ref) {
     }
   };
 
+  const handleEntered = (element, isAppearing) => {
+    if (atLeastOneSubMenu) setEntering(false);
+
+    if (onEntered) {
+      onEntered(element, isAppearing);
+    }
+  };
+
+  const handleOnClose = (event) => {
+    event.preventDefault();
+    setOpenSubMenuIndex(null);
+    if (onClose) {
+      onClose(event, `${event.key.toLowerCase()}KeyDown`);
+    }
+  };
+
   const handleListKeyDown = (event) => {
-    if (event.key === 'Tab') {
-      event.preventDefault();
+    if (event.key === 'Tab' || event.key === 'Escape') {
+      handleOnClose(event);
+    }
 
-      if (onClose) {
-        onClose(event, 'tabKeyDown');
-      }
+    if (event.key === 'ArrowLeft' && isSubMenu) {
+      // Tell the parent Menu to close the sub Menu that you're in, but
+      // don't trigger the sub Menu onClose cascade.
+      event.stopPropagation();
```

Comment:
Same concern here: Listen to event.defaultPrevented in the parent should be enough.

Reply:
I think we need to do this because of the recursive structure of the cascading menu and the use of the same components at every level. Because this is the same `Menu` component at the level where this event is going to originate as the parent level where I may be able to listen for things like event.defaultPrevented, it has to actually stop the event at some point.

So it seems like every level needs to be able to actively stop the events from bubbling.

This is the opposite of the `onArrowRightKeydown` operation in that it sets the parent Menu's openSubMenuIndex back to null, which closes the child subMenu where this event originates. I only want the current menu to close when I arrow left, so the event must be isolated to the originating node.

Reply:
> I think we need to do this because of the recursive structure of the cascading menu and the use of the same components at every level. Because this is the same Menu component at the level where this event is going to originate as the parent level where I may be able to listen for things like event.defaultPrevented, it has to actually stop the event at some point.

Could you explain why a component above can't ignore the event if defaultPrevented? 

I don't see how 

|- A: if (!event.defaultPrevented) handleEvent()
|-- B: event.preventDefault()

is different than 
|- A
|-- B: event.stopPropagation()

Reply:
Those look equivalent, I think, so I'll give that a try and make sure it still works as it should. This is a bit of a weak area for me, I think, so I may have misunderstood what you meant.

I did a little digging and there is a (seemingly) pretty good [css-tricks article](https://css-tricks.com/dangers-stopping-event-propagation/) about avoiding the use of `event.stopPropagation` wherever you can. Is this the primary set of concerns you're coming from? This is news to me, so I'd like to seize the opportunity to learn something that seems fairly important.

Reply:
> event.stopPropagation wherever you can. Is this the primary set of concerns you're coming from? 

My opinion on that is not evidence or use-case based. It just irks me when I see this since it's essentially a component changing the global flow of events. The problem is that once the propagation is stopped, it's stopped. If some app needs to listen to bubbled keydown events they are shut off. But with preventDefault we can keep the event flow and tell that it was already handled.

Reply:
Cool. That's more or less what the css-tricks article (and a few others) were saying. I've never been bitten in this way, so it never occurred to me to rethink that. Thanks for pointing it out. This is good stuff.

Reply:
Pushed a fix.

Reply:
> That's more or less what the css-tricks article (and a few others) were saying. I've never been bitten in this way, so it never occurred to me to rethink that.

Me neither and it's possible that this isn't actually a concern. It just makes more sense to me at a high-level to say "I've handled this" instead of "you don't get to handle this!". It's also motivated by removing potential surprise when events suddenly stop bubbling up. That can be especially frustrating with 3rd party components.

Reply:
Yeah, I've certainly changed my approach here and will follow this approach going forward.